apple-dmg/src/lib.rs

```diff
@@ -0,0 +1,346 @@
+            data_hasher: Hasher::new(),
+            main_hasher: Hasher::new(),
+            sector_number: 0,
+            compressed_offset: 0,
+        }
+    }
+
+    pub fn create_fat32(mut self, fat32: &[u8]) -> Result<()> {
+        anyhow::ensure!(fat32.len() % 512 == 0);
+        let sector_count = fat32.len() as u64 / 512;
+        let mut mbr = ProtectiveMBR::new();
+        let mut partition = PartRecord::new_protective(Some(sector_count.try_into()?));
+        partition.os_type = 11;
+        mbr.set_partition(0, partition);
+        let mbr = mbr.as_bytes()?;
+        self.add_partition("Master Boot Record (MBR : 0)", &mbr)?;
+        self.add_partition("FAT32 (FAT32 : 1)", fat32)?;
+        self.finish()?;
+        Ok(())
+    }
+
+    pub fn add_partition(&mut self, name: &str, bytes: &[u8]) -> Result<()> {
+        anyhow::ensure!(bytes.len() % 512 == 0);
+        let id = self.xml.partitions().len() as u32;
+        let name = name.to_string();
+        let mut table = BlkxTable::new(id, self.sector_number, crc32fast::hash(bytes));
+        for chunk in bytes.chunks(2048 * 512) {
+            let mut encoder = ZlibEncoder::new(chunk, Compression::best());
+            let mut compressed = vec![];
+            encoder.read_to_end(&mut compressed)?;
+            let compressed_length = compressed.len() as u64;
+            let sector_count = chunk.len() as u64 / 512;
+            self.w.write_all(&compressed)?;
+            self.data_hasher.update(&compressed);
+            table.add_chunk(BlkxChunk::new(
+                ChunkType::Zlib,
+                self.sector_number,
+                sector_count,
+                self.compressed_offset,
+                compressed_length,
+            ));
+            self.sector_number += sector_count;
+            self.compressed_offset += compressed_length;
+        }
+        table.add_chunk(BlkxChunk::term(self.sector_number, self.compressed_offset));
+        self.main_hasher.update(&table.checksum.data[..4]);
+        self.xml
+            .add_partition(Partition::new(id as i32 - 1, name, table));
+        Ok(())
+    }
+
+    pub fn finish(mut self) -> Result<()> {
+        let mut xml = vec![];
+        plist::to_writer_xml(&mut xml, &self.xml)?;
+        let pos = self.w.seek(SeekFrom::Current(0))?;
+        let data_digest = self.data_hasher.finalize();
+        let main_digest = self.main_hasher.finalize();
+        let koly = KolyTrailer::new(
+            pos,
+            self.sector_number,
+            pos,
+            xml.len() as _,
+            data_digest,
+            main_digest,
+        );
+        self.w.write_all(&xml)?;
+        koly.write_to(&mut self.w)?;
+        Ok(())
+    }
+}
+
+fn symlink(target: &str) -> Vec<u8> {
+    let xsym = format!(
+        "XSym\n{:04}\n{:x}\n{}\n",
+        target.as_bytes().len(),
+        md5::compute(target.as_bytes()),
+        target,
+    );
+    let mut xsym = xsym.into_bytes();
+    xsym.resize(1067, b' ');
```

Comment:
Where does `1067` come from?

I think there's a bug in here where the formatting of `target` is > 1067 bytes? Should we check for that and make this function return `Result`?